crates/ruff_python_semantic/src/analyze/type_inference.rs

```diff
@@ -24,22 +24,38 @@ impl ResolvedPythonType {
             (Self::TypeError, _) | (_, Self::TypeError) => Self::TypeError,
             (Self::Unknown, _) | (_, Self::Unknown) => Self::Unknown,
             (Self::Atom(a), Self::Atom(b)) => {
-                if a == b {
+                if a.is_subtype_of(b) {
+                    Self::Atom(b)
+                } else if b.is_subtype_of(a) {
                     Self::Atom(a)
                 } else {
                     Self::Union(FxHashSet::from_iter([a, b]))
                 }
             }
             (Self::Atom(a), Self::Union(mut b)) => {
-                b.insert(a);
+                // If `a` is a subtype of any of the types in `b`, then `a` is
+                // redundant.
+                if !b.iter().any(|type_| a.is_subtype_of(*type_)) {
+                    b.insert(a);
+                }
                 Self::Union(b)
             }
             (Self::Union(mut a), Self::Atom(b)) => {
-                a.insert(b);
+                // If `b` is a subtype of any of the types in `a`, then `b` is
+                // redundant.
+                if !a.iter().any(|type_| b.is_subtype_of(*type_)) {
+                    a.insert(b);
+                }
                 Self::Union(a)
             }
             (Self::Union(mut a), Self::Union(b)) => {
-                a.extend(b);
+                for type_ in b {
+                    // If `type_` is a subtype of any of the types in `a`, then
+                    // `type_` is redundant.
+                    if !a.iter().any(|type_| type_.is_subtype_of(*type_)) {
+                        a.insert(type_);
+                    }
+                }
```

Comment:
I think we would need to have different variable names here for `type_` like `type_a`, `type_b`.